File: 290. Word Pattern/solution.cpp
class Solution {
public:
  bool wordPattern(string pattern, string s) {
    vector<string> sList;
    string word = "";
    for (int index = 0; index < s.size(); index++){
      char letter = s[index];
      if (letter != ' '){
        word += letter;
      }
      if (letter == ' ' || index == s.size() - 1){
        sList.push_back(word);
        word = "";
      }
    }
    
    if (sList.size() != pattern.size()){
      return false;
    }

    unordered_map<string, string> patternDictionary;
    unordered_map<string, string> wordDictionary;

    for (int index = 0; index < sList.size(); index++){
      string word = sList[index];
      string letter (1, pattern[index]);
      if (patternDictionary.find(letter) != patternDictionary.end()){
        if (patternDictionary[letter] != word){
          return false;
        }
      }
      else{
        if (wordDictionary.find(word) != wordDictionary.end()){
          if (wordDictionary[word] != letter){
            return false;
          }
        }
        patternDictionary[letter] = word;
        wordDictionary[word] = letter;
      }
    }

    return true;
  }
};

File: 290. Word Pattern/solution.js
/**
 * @param {string} pattern
 * @param {string} s
 * @return {boolean}
 */
 var wordPattern = function(pattern, s) {
  let sList = s.split(" ");
  if (sList.length != pattern.length){
    return false;    
  }

  let patternDictionary = {};
  let wordDictionary = {};

  for (let index = 0; index < sList.length; index++){
    let word = sList[index];
    let letter = pattern[index];
    if (patternDictionary[letter] != undefined){
      if (patternDictionary[letter] != word){
        return false;
      }
    }
    else{
      if (wordDictionary[word] != undefined){
        if (wordDictionary[word] != letter){
          return false;
        }
      }
      patternDictionary[letter] = word;
      wordDictionary[word] = letter;
    }
  }

  return true;
};